### PR TITLE
feat(protect): add support for deeply nested protect schemas

### PR DESCRIPTION
**Pull request by calvinbrewer**
Added support for nested objects in Protect schemas, allowing you to encrypt and search on nested properties through the model interfaces. E.g.

```ts
import { csTable, csColumn } from "@cipherstash/protect";

export const protectedUsers = csTable("users", {
  email: csColumn("email").freeTextSearch().equality().orderAndRange(),
  profile: {
    name: csColumn("name").freeTextSearch(),
    address: {
      street: csColumn("street").freeTextSearch(),
      location: {
        coordinates: csColumn("coordinates").equality(),
      },
    },
  },
});
```

When working with nested objects:
- Each level can have its own encrypted fields
- Index options can be applied to any level of nesting
- The maximum nesting depth is 3 levels
- Null and undefined values are supported at any level
- Optional nested objects are supported

### CODE REVIEW

File: packages/protect/src/ffi/index.ts [hunk 1]
```diff
@@ -68,6 +68,8 @@ export class ProtectClient {
           client_key: config.clientKey,
         })
 
+        console.log('config', validated)
```
Comment (CDThomas) [inline packages/protect/src/ffi/index.ts]: Heads up that a `console.log` got left behind here:
```suggestion
```

File: packages/protect/__tests__/nested-models.test.ts [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+  address?: string | null
+  notEncrypted?: string | null
+  example: {
+    field: string | undefined | null
+    nested?: {
+      deeper: string | undefined | null
+    }
+  }
+}
+
```
Comment (CDThomas) [inline packages/protect/__tests__/nested-models.test.ts:45]: I think it'd be worth adding some tests for:
- a mix of plaintext and encrypted nested fields
- partial matches of paths

File: packages/protect/src/ffi/model-helpers.ts [hunk 3]
```diff
@@ -378,18 +499,40 @@ export async function bulkEncryptModels<T extends Record<string, unknown>>(
...
     keyMap,
   )
 
-  // Reconstruct models
-  return models.map((_, modelIndex) => ({
-    ...otherFields[modelIndex],
-    ...nullFields[modelIndex],
-    ...Object.fromEntries(
+  // Helper function to set a nested value
+  const setNestedValue = (
```
Comment (CDThomas) [inline packages/protect/src/ffi/model-helpers.ts:655]: Is this the same as the `setNestedValue` function above? If yes, it might be worth extracting this function and sharing it.

File: packages/protect/src/ffi/model-helpers.ts [hunk 4]
```diff
@@ -421,26 +571,47 @@ export async function bulkDecryptModels<T extends Record<string, unknown>>(
...
```
Comment (CDThomas) [inline packages/protect/src/ffi/model-helpers.ts:734]: Same comment here on removing the duplicate `setNestedValue` functions.

File: docs/reference/schema.md [hunk 5]
```diff
@@ -75,6 +76,45 @@ export const protectedUsers = csTable("users", {
 });
 ```
 
+### Nested objects
+
+Protect.js supports nested objects in your schema, allowing you to encrypt **but not search on** nested properties. You can define nested objects up to 3 levels deep.
+This is useful for data stores that have less structured data, like NoSQL databases.
+
+You can define nested objects by using the `csValue` function to define a value in a nested object. The value naming convention of the `csValue` function is a dot-separated string of the nested object path, e.g. `profile.name` or `profile.address.street`.
```
Comment (CDThomas) [inline docs/reference/schema.md:84]: The column/attribute name in the DB table can't contain a `.` since we're using it a delimiter, right? Definitely an edge case, but thought it was worth mentioning.

**Review by CDThomas — APPROVED** [review]
Left a few minor comments, but LGTM otherwise.

File: packages/protect/__tests__/nested-models.test.ts (hunk 2 above)
Comment (calvinbrewer) [reply packages/protect/__tests__/nested-models.test.ts:45]: Great call @CDThomas ran into a bunch of issues when adding those tests that I didn't consider. Added fixes for them and the addtional tests

File: packages/protect/src/ffi/model-helpers.ts (hunk 3 above)
Comment (calvinbrewer) [reply packages/protect/src/ffi/model-helpers.ts:655]: After Protect.js is a bit more streamlined - we should definitely do some refactoring exercises

File: packages/protect/src/ffi/model-helpers.ts (hunk 4 above)
Comment (calvinbrewer) [reply packages/protect/src/ffi/model-helpers.ts:734]: After Protect.js is a bit more streamlined - we should definitely do some refactoring exercises
